Reuse WebServicesProxy object when getting solutions publish messages

sergiomiranda-outsystems opened the pull request:
Reuse WebServicesProxy object when getting solutions publish messages to avoid constantly recreating it, which would cause the module to gradually increase memory utilization during execution. When publishing a large number of modules this would make the function eventually fail with an OutOfMemory error.

In my tests publishing 364 modules, the function would reach about 4GB of memory use and then fail with the same OutOfMEmory error reported in issue #102.
After this change, the same solution publish finished successfully and the memory use of powershell.exe was constant and always around 100MB.

sergiomiranda-outsystems committed:
Reuse WebServicesProxy object when getting solutions publish messages

Reuse WebServicesProxy object when getting solutions publish messages to avoid constantly recreating it, which would cause the module to gradually increase memory utilization during execution.
When publishing a large number of modules this would make the function eventually fail with an OutOfMemory error.

pedro132 reviewed (commented):
A couple of comments:
* I get that function _SCWS_SolutionPack_GetPublicationMessages_ is the function that's called more often and so where most of new objects were getting created without need. Still, the same no-reuse logic happens in other places that _SCWS_GetPlatformServicesProxy_ is used; and for that matter, for other places in the same file, where we are creating SOAP clients.
  * Can you move the fix to SOAP getter method (_SCWS_GetPlatformServicesProxy_ ) and apply it to the other SOAP consumers please 🙏 
* If possible, include QA

luisalourenco committed:
Merge branch 'dev' into Publish-OSPlatformModule

sergiomiranda-outsystems committed:
Move object reuse logic and replicate it to other functions

Move object reuse logic to the function where the WebServiceProxy object is actually created, and replicate it to other similar functions for other SOAP endpoints.

sergiomiranda-outsystems committed:
Small fix to wrong variable used

luisalourenco committed:
Merge branch 'dev' into Publish-OSPlatformModule

joaorosado reviewed (commented):
I don't think this issue will be easy to fix like this.
While it does mitigate the problem it does feels a lot like a workaround that deserves a different approach.

In particular all the methods should actually instead receive the ws proxy and the caller should be the one responsible for managing it's lifecyle (and even do things like Close or Dispose of the objects they create).

I also noticed a lot of more problems in this file, like never using https endpoints for any communications.
So this topic will probably require some more attention from our team to rethink than just pinpoint changes.

sergiomiranda-outsystems commented:
Hello @joaorosado . Thank you for your insights.
I understand that there might be a better way to solve this, but my focus here was mosly to fix the OutOfMemory error and get the function to work with large solutions without rewriting it extensively.

About HTTPS not being used/supported, there is an old [issue ](https://github.com/OutSystems/OutSystems.SetupTools/issues/82) open about that and I was already doing some work about it, but if your team will work on that as well I will leave it.

pedro132 reviewed (approved):
> Hello @joaorosado . Thank you for your insights. I understand that there might be a better way to solve this, but my focus here was mosly to fix the OutOfMemory error and get the function to work with large solutions without rewriting it extensively. 

My two cents: agreed that a fix to stop the pathological memory leak (which is an old problem) shouldn't stop us from doing a more full fix. Conversely, this PR has plenty of value by itself so it should be merged IMO.

luisalourenco commented:
Merge on hold until we release a new SetupTools with the FIPS changes included and drop support for O10
